kernel/rros/flags.rs

```diff
@@ -66,6 +75,16 @@ impl RrosFlag {
         }
     }
 
+    #[inline]
+    pub fn wait_same(&mut self) -> i32 {
+        // TODO: Try to get around the limitations of immutable borrowing.
+        let mut x = unsafe { NonNull::new_unchecked(&self.wait as *const _ as *mut RrosWaitQueue) };
```

Comment:
Can't we directly call `self.wait.wait_timeout(...)`?

Reply:
We can't use it directly like this. Because the conditional closure function needs to hold an immutable reference to self at the same time. But `Rust` does not allow having both mutable and immutable references at the same time.
```Rust
    #[inline]
    pub fn wait_same(&mut self) -> i32 {
        self.wait.wait_timeout(
            RROS_INFINITE,
            RrosTmode::RrosRel,
            || self.peek(),
        )
    }
```

Reply:
If `self.peak` is mutable, how can you guarantee the list is checked atomically?

Reply:
If so, this should be implemented with a `sync` trait. It seems that you are sending a reference to another thread. With a raw pointer, you can pass compiler but the safety can not be assured.

Reply:
No, I think I didn't explained it clearly.

The problem here is `Rust` can't have mutable and immutable references at the same time. The compiler error is:

![image](https://github.com/user-attachments/assets/357b5497-d200-4719-911a-9cca760b80ed)

However, I found a solution here: [Splitting Borrows](https://doc.rust-lang.org/nomicon/borrow-splitting.html). This code can be compiled successfully by changing it to this:

```Rust
    #[inline]
    pub fn wait_same(&mut self) -> i32 {
        let cell = &self.raised;

        self.wait.wait_timeout(
            RROS_INFINITE,
            RrosTmode::RrosRel,
            || {
                unsafe { rust_helper_smp_wmb(); }
                cell.get()
            },
        )
    }
```